File: Problems/Algorithms/473. Matchsticks to Square/matchsticks.go
package main

import (
    "sort"
)

func dfs(nums []int, sums []int, length int, index int) bool {
    if index == len(nums) {
        return sums[0] == sums[1] && sums[0] == sums[2] && sums[0] == sums[3]
    }

    curr := nums[index]
    if curr > length {
        return false
    }

    for i := range sums {
        if curr + sums[i] <= length {
            sums[i] += curr
            if dfs(nums, sums, length, index+1) {
                return true
            }
            sums[i] -= curr
        }
    }

    return false
}

func makesquare(matchsticks []int) bool {
    if matchsticks == nil || len(matchsticks) == 0 {
        return false
    }

    sort.Sort(sort.Reverse(sort.IntSlice(matchsticks)))

    perimeter := 0
    for i := range matchsticks {
        perimeter += matchsticks[i]
    }
    if perimeter % 4 != 0 {
        return false
    }
    length := perimeter / 4

    sums := make([]int, 4)

    return dfs(matchsticks, sums, length, 0)
}

File: Problems/Algorithms/473. Matchsticks to Square/matchsticks.py
class Solution:
    def makesquare(self, matchsticks: List[int]) -> bool:
        if not matchsticks:
            return False

        perimeter = sum(matchsticks)
        if perimeter % 4 != 0:
            return False
        side = perimeter // 4

        memo = {}
        n = len(matchsticks)

        def aux(mask, sides):
            used = 0
            for i in range(n):
                if not (mask & (1 << i)):
                    used += matchsticks[i]

            if used > 0 and used % side == 0:
                sides += 1

            if sides == 3:
                return True

            if (mask, sides) in memo:
                return memo[(mask, sides)]

            ans = False
            rest = side
            if used % side != 0:
                rest = side - used % side

            for i in range(n):
                if matchsticks[i] <= rest and mask & (1 << i):
                    if aux(mask ^ (1 << i), sides):
                        ans = True
                        break

            memo[(mask, sides)] = ans
            return ans

        return aux((1 << n) - 1, 0)

